internal/consul/connection.go

```diff
@@ -1,50 +1,192 @@
 package consul
 
 import (
+	"context"
+	"crypto/tls"
+	"os"
+	"strings"
+	"sync"
+
+	"errors"
+	"fmt"
+	"github.com/hashicorp/consul-server-connection-manager/discovery"
 	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
 )
 
+const ConsulHTTPAddressEnvName = "CONSUL_HTTP_ADDR"
+const DiscoveryKey = "exec=discover"
+
 type Client interface {
 	Agent() *api.Agent
 	Catalog() *api.Catalog
 	ConfigEntries() *api.ConfigEntries
 	DiscoveryChain() *api.DiscoveryChain
 	Namespaces() *api.Namespaces
 
+	Watch(ctx context.Context) error
+
+	Token() string
+
 	// TODO: drop this
 	Internal() *api.Client
 }
 
+type ClientConfig struct {
+	Addresses   string
+	HTTPAddress string
+	HTTPPort    int
+	GRPCPort    int
+	Namespace   string
+	TLS         *tls.Config
+	Credentials discovery.Credentials
+	Logger      hclog.Logger
+}
+
 type client struct {
-	client *api.Client
+	config      ClientConfig
+	client      *api.Client
+	token       string
+	mutex       sync.RWMutex
+	initialized chan struct{}
 }
 
-func NewClient(c *api.Client) Client {
+func NewClient(config ClientConfig) Client {
 	return &client{
-		client: c,
+		config:      config,
+		initialized: make(chan struct{}),
+	}
+}
+
+func (c *client) wait() {
+	<-c.initialized
+}
+
+func ParseDiscoveryAddresses() (addresses string, err error) {
+	//TODO should this be pulled in from the flag instead of the env?
+	consulhttpAddress := os.Getenv(ConsulHTTPAddressEnvName)
+	if !strings.Contains(consulhttpAddress, DiscoveryKey) {
+		//TODO should this return an error? What do we do in this case? Default?
+		return "", errors.New("discovery not found")
+	}
+	s := strings.Split(consulhttpAddress, ":")
+	return s[0], nil
+
+}
+
+func (c *client) Watch(ctx context.Context) error {
+	watcher, err := discovery.NewWatcher(ctx, discovery.Config{
+		Addresses:   c.config.Addresses,
+		GRPCPort:    c.config.GRPCPort,
+		TLS:         c.config.TLS,
+		Credentials: c.config.Credentials,
+	}, c.config.Logger)
+	if err != nil {
+		return err
+	}
+	go watcher.Run()
+	defer watcher.Stop()
+
+	// Wait for initial state.
+	state, err := watcher.State()
+	if err != nil {
+		return err
+	}
+	updateClient := func(s discovery.State) error {
+		cfg := api.DefaultConfig()
```

Comment:
Since we do some additional TLS overriding in our entrypoint, I think we'll need to do that here too. We probably want to pass the base client config for Consul in via `ClientConfig` and then just explicitly override the address and token based on the discovery state.

Reply:
Do my updates match more what you had in mind?